salt/utils/master.py

```diff
@@ -39,6 +39,114 @@
 log = logging.getLogger(__name__)
 
 
+def running(opts):
+    '''
+    Return the running jobs on this minion
+    '''
+
+    ret = []
+    proc_dir = os.path.join(opts['cachedir'], 'proc')
+    if not os.path.isdir(proc_dir):
+        return ret
+    for fn_ in os.listdir(proc_dir):
+        path = os.path.join(proc_dir, fn_)
+        try:
+            data = _read_proc_file(path, opts)
+            if data is not None:
+                ret.append(data)
+        except (IOError, OSError):
+            # proc files may be removed at any time during this process by
+            # the minion process that is executing the JID in question, so
+            # we must ignore ENOENT during this process
+            pass
+    return ret
+
+
+def cache_jobs(opts, jid, ret):
+    serial = salt.payload.Serial(opts=opts)
+
+    fn_ = os.path.join(
+        opts['cachedir'],
+        'minion_jobs',
+        jid,
+        'return.p')
+    jdir = os.path.dirname(fn_)
+    if not os.path.isdir(jdir):
+        os.makedirs(jdir)
+    with salt.utils.files.fopen(fn_, 'w+b') as fp_:
+        fp_.write(serial.dumps(ret))
+
+
+def _read_proc_file(path, opts):
+    '''
+    Return a dict of JID metadata, or None
+    '''
+    serial = salt.payload.Serial(opts)
+    with salt.utils.files.fopen(path, 'rb') as fp_:
+        buf = fp_.read()
+        fp_.close()
+        if buf:
+            data = serial.loads(buf)
+        else:
+            # Proc file is empty, remove
+            try:
+                os.remove(path)
+            except IOError:
+                pass
+            return None
+    if not isinstance(data, dict):
+        # Invalid serial object
+        return None
+    if not salt.utils.process.os_is_running(data['pid']):
+        # The process is no longer running, clear out the file and
+        # continue
+        try:
+            os.remove(path)
+        except IOError:
+            pass
+        return None
+
+    if not _check_cmdline(data):
+        pid = data.get('pid')
+        if pid:
+            log.warning(
+                'PID %s exists but does not appear to be a salt process.', pid
+            )
+        try:
+            os.remove(path)
+        except IOError:
+            pass
+        return None
```

Comment:
This is where it gets scattered. You are agree with cmdline for _any_ command that contains `salt` string, as well as it is not even Linux. This should not be done that way. This should say `True` only if the command line really is Salt command. But then you cannot guarantee it either, because you're banning all OSes that aren't Linux but still implements procfs.

Reply:
This is code that is only used by the Salt master, which is only supported on Linux based operating systems.  If/when that changes in the future, we'll definitely revisit this code so it's supported on other OSes.